bundles/org.eclipse.team.core/src/org/eclipse/team/internal/core/TeamHookDispatcher.java

```diff
@@ -31,13 +35,17 @@
  */
 public class TeamHookDispatcher extends TeamHook {
 
-	private static final ResourceRuleFactory defaultFactory = new ResourceRuleFactory() {};
+	private static final Map<IWorkspace, IResourceRuleFactory> defaultFactoryMap = new ConcurrentHashMap<>();
```

Comment:
Why the map here? Shouldn't the TeamHook be workspace-specific already?

Reply:
Theoretically one could call the method with projects from different workspaces and as one has to store the object somewhere anyways I just used a map here.

Reply:
No, please don't over-engineer. We have **single** workspace instance. Multiple workspaces will be in Eclipse 5.x if ever.

Reply:
I vote for over-engineering too here; and what you call defensive programming IMO doesn't apply as it seems like the TeamHook has to be a unique for a workspace. The code here would allow to have a single static service for all workspaces, which wouldn't really make future code better. Moreover this code is harder to maintain and consumes more CPU/RAM. Please stick to some code assuming there is a single workspace.

Reply:
The last "maintain" of this code was 16 years ago...

Reply:
> The last "maintain" of this code was 16 years ago...

Not sure what you wanted to say with that. I guess respect for the stable code base with no bugs, otherwise it would be constantly modified.

Reply:
> Not sure what you wanted to say with that.
> I guess respect for the stable code base with no bugs

Yes, this code seem not needing much maintain, so even if it would be a little harder with new code but won't need to be touched for other 16 years I think "harder to maintain" is not really an issue here...